dali/util/uri.h

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_UTIL_URI_H_
+#define DALI_UTIL_URI_H_
+
+#include <cassert>
+#include <cstring>
+#include <stdexcept>
+#include <string>
+#include <string_view>
+#include "dali/core/api_helper.h"
+
+namespace dali {
+
+class DLL_PUBLIC URI {
+ private:
+  std::string uri_;  // the original URI string
+  std::ptrdiff_t scheme_start_ = 0, scheme_end_ = 0;
+  std::ptrdiff_t authority_start_ = 0, authority_end_ = 0;
+  std::ptrdiff_t path_start_ = 0, path_end_ = 0;
+  std::ptrdiff_t query_start_ = 0, query_end_ = 0;
+  std::ptrdiff_t fragment_start_ = 0, fragment_end_ = 0;
+  bool valid_ = true;
```

Comment:
Empty URI is invalid (no scheme)
```suggestion
  bool valid_ = false;
```